{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "88ad2789",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import torch.nn as nn\n",
    "import torch\n",
    "cos = nn.CosineSimilarity(dim=0, eps=1e-6)\n",
    "import torch.nn.functional as F\n",
    "from Bio import AlignIO\n",
    "import pandas as pd\n",
    "import torch\n",
    "import esm\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from Bio.PDB import PDBParser, PPBuilder\n",
    "import warnings\n",
    "from Bio.PDB.PDBExceptions import PDBConstructionWarning\n",
    "from Bio.SeqUtils import IUPACData\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad0be7e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = '/Users/williamharrigan/Desktop/Github/contact_site_classifier/attention_classifier/data_files/'\n",
    "fasta_file = base_dir + 'rcsb_pdb_3KYN.fasta'\n",
    "pdb_filename = base_dir + '3kyn.pdb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "be9df12d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## This code is not important right now, will be when adding multiple sequences\n",
    "\n",
    "# seqs = SeqIO.to_dict(SeqIO.parse(full_len_sequences, \"fasta\"))\n",
    "# for k,v in seqs.items():\n",
    "#     seqs[k] = str(v.seq)\n",
    "    \n",
    "# print(list(seqs.keys())[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0c7f3b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simple_aa(three_letter_code):\n",
    "    return IUPACData.protein_letters_3to1.get(three_letter_code.capitalize())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4810529d",
   "metadata": {},
   "source": [
    "## Generate ESM-2 Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cf774bbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ESM2(\n",
       "  (embed_tokens): Embedding(33, 1280, padding_idx=1)\n",
       "  (layers): ModuleList(\n",
       "    (0-32): 33 x TransformerLayer(\n",
       "      (self_attn): MultiheadAttention(\n",
       "        (k_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
       "        (v_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
       "        (q_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
       "        (out_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
       "        (rot_emb): RotaryEmbedding()\n",
       "      )\n",
       "      (self_attn_layer_norm): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
       "      (fc1): Linear(in_features=1280, out_features=5120, bias=True)\n",
       "      (fc2): Linear(in_features=5120, out_features=1280, bias=True)\n",
       "      (final_layer_norm): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
       "    )\n",
       "  )\n",
       "  (contact_head): ContactPredictionHead(\n",
       "    (regression): Linear(in_features=660, out_features=1, bias=True)\n",
       "    (activation): Sigmoid()\n",
       "  )\n",
       "  (emb_layer_norm_after): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
       "  (lm_head): RobertaLMHead(\n",
       "    (dense): Linear(in_features=1280, out_features=1280, bias=True)\n",
       "    (layer_norm): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Load the ESM Model\n",
    "model, alphabet = esm.pretrained.esm2_t33_650M_UR50D()\n",
    "model.eval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "77eaf392",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set model to use cuda GPU\n",
    "if torch.cuda.is_available():\n",
    "    model = model.cuda()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b65b63ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3KYN\n",
      "Sequence:  MSHSMRYFSAAVSRP \n",
      "\n",
      "Data:  [(0, 'MSHSMRYFSAAVSRPGRGEPRFIAMGYVDDTQFVRFDSDSASPRMEPRAPWVEQEGPEYWEEETRNTKAHAQTDRMNLQTLRGYYNQSEASSHTLQWMIGCDLGSDGRLIRGYERYAYDGKDYLALNEDLRSWTAADTAAQISKRKCEAANVAEQRRAYLEGTCVEWLHRYLENGKEMLQRADPPKTHVTHHPVFDYEATLRCWALGFYPAEIILTWQRDGEDQTQDVELVETRPAGDGTFQKWAAVVVPSGEEQRYTCHVQHEGLPEPLMLRWKMIQRTPKIQVYSRHPAENGKSNFLNCYVSGFHPSDIEVDLLKNGERIEKVEHSDLSFSKDWSFYLLYYTEFTPTEKDEYACRVNHVTLSQPKIVKWDRDMKGPPAALTL')] \n",
      "\n",
      "batch_tokens:  \n",
      "\n",
      " tensor([[ 0, 20,  8, 21,  8, 20, 10, 19, 18,  8,  5,  5,  7,  8, 10, 14,  6, 10,\n",
      "          6,  9, 14, 10, 18, 12,  5, 20,  6, 19,  7, 13, 13, 11, 16, 18,  7, 10,\n",
      "         18, 13,  8, 13,  8,  5,  8, 14, 10, 20,  9, 14, 10,  5, 14, 22,  7,  9,\n",
      "         16,  9,  6, 14,  9, 19, 22,  9,  9,  9, 11, 10, 17, 11, 15,  5, 21,  5,\n",
      "         16, 11, 13, 10, 20, 17,  4, 16, 11,  4, 10,  6, 19, 19, 17, 16,  8,  9,\n",
      "          5,  8,  8, 21, 11,  4, 16, 22, 20, 12,  6, 23, 13,  4,  6,  8, 13,  6,\n",
      "         10,  4, 12, 10,  6, 19,  9, 10, 19,  5, 19, 13,  6, 15, 13, 19,  4,  5,\n",
      "          4, 17,  9, 13,  4, 10,  8, 22, 11,  5,  5, 13, 11,  5,  5, 16, 12,  8,\n",
      "         15, 10, 15, 23,  9,  5,  5, 17,  7,  5,  9, 16, 10, 10,  5, 19,  4,  9,\n",
      "          6, 11, 23,  7,  9, 22,  4, 21, 10, 19,  4,  9, 17,  6, 15,  9, 20,  4,\n",
      "         16, 10,  5, 13, 14, 14, 15, 11, 21,  7, 11, 21, 21, 14,  7, 18, 13, 19,\n",
      "          9,  5, 11,  4, 10, 23, 22,  5,  4,  6, 18, 19, 14,  5,  9, 12, 12,  4,\n",
      "         11, 22, 16, 10, 13,  6,  9, 13, 16, 11, 16, 13,  7,  9,  4,  7,  9, 11,\n",
      "         10, 14,  5,  6, 13,  6, 11, 18, 16, 15, 22,  5,  5,  7,  7,  7, 14,  8,\n",
      "          6,  9,  9, 16, 10, 19, 11, 23, 21,  7, 16, 21,  9,  6,  4, 14,  9, 14,\n",
      "          4, 20,  4, 10, 22, 15, 20, 12, 16, 10, 11, 14, 15, 12, 16,  7, 19,  8,\n",
      "         10, 21, 14,  5,  9, 17,  6, 15,  8, 17, 18,  4, 17, 23, 19,  7,  8,  6,\n",
      "         18, 21, 14,  8, 13, 12,  9,  7, 13,  4,  4, 15, 17,  6,  9, 10, 12,  9,\n",
      "         15,  7,  9, 21,  8, 13,  4,  8, 18,  8, 15, 13, 22,  8, 18, 19,  4,  4,\n",
      "         19, 19, 11,  9, 18, 11, 14, 11,  9, 15, 13,  9, 19,  5, 23, 10,  7, 17,\n",
      "         21,  7, 11,  4,  8, 16, 14, 15, 12,  7, 15, 22, 13, 10, 13, 20, 15,  6,\n",
      "         14, 14,  5,  5,  4, 11,  4,  2]])\n"
     ]
    }
   ],
   "source": [
    "# 2. Prepare sequence input\n",
    "# Load sequence file\n",
    "\n",
    "seq_chains = []\n",
    "\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    seq_chains.append(str(record.seq))\n",
    "\n",
    "pdb_id = record.id.split('|')[0].split('_')[0]\n",
    "protein_sequence = ''.join(seq_chains)\n",
    "print(pdb_id)\n",
    "print('Sequence: ', protein_sequence[:15], '\\n')\n",
    "\n",
    "# Index protein sequence as sequence 0 (next sequence would be indexed as 1)\n",
    "data = [(0, protein_sequence)]\n",
    "print('Data: ', data, '\\n')\n",
    "\n",
    "# Prepare variables to input sequence into ESM-2 model \n",
    "batch_converter = alphabet.get_batch_converter()\n",
    "batch_labels, batch_strs, batch_tokens = batch_converter(data)\n",
    "batch_tokens = batch_tokens.cuda() if torch.cuda.is_available() else batch_tokens\n",
    "\n",
    "print('batch_tokens: ', '\\n\\n', batch_tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "acf92a62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model outputs:  dict_keys(['logits', 'representations', 'attentions', 'contacts'])\n"
     ]
    }
   ],
   "source": [
    "# 4. Input prepared sequence information into model and output as results (contact predictions are included in embedding)\n",
    "with torch.no_grad():\n",
    "    results = model(batch_tokens, repr_layers=[33], return_contacts=True)\n",
    "\n",
    "print('Model outputs: ', results.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32f5645e",
   "metadata": {},
   "source": [
    "## Double Checking Indexes for Embedding Sequence vs PDB Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "184651c5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Chain id=A>\n",
      "0\n",
      "1 M M <Atom CA>\n",
      "2 S S <Atom CA>\n",
      "3 H H <Atom CA>\n",
      "4 S S <Atom CA>\n",
      "5 M M <Atom CA>\n",
      "6 R R <Atom CA>\n",
      "7 Y Y <Atom CA>\n",
      "8 F F <Atom CA>\n",
      "9 S S <Atom CA>\n",
      "10 A A <Atom CA>\n",
      "11 A A <Atom CA>\n",
      "12 V V <Atom CA>\n",
      "13 S S <Atom CA>\n",
      "14 R R <Atom CA>\n",
      "15 P P <Atom CA>\n",
      "16 G G <Atom CA>\n",
      "17 R R <Atom CA>\n",
      "18 G G <Atom CA>\n",
      "19 E E <Atom CA>\n",
      "20 P P <Atom CA>\n",
      "21 R R <Atom CA>\n",
      "22 F F <Atom CA>\n",
      "23 I I <Atom CA>\n",
      "24 A A <Atom CA>\n",
      "25 M M <Atom CA>\n",
      "26 G G <Atom CA>\n",
      "27 Y Y <Atom CA>\n",
      "28 V V <Atom CA>\n",
      "29 D D <Atom CA>\n",
      "30 D D <Atom CA>\n",
      "31 T T <Atom CA>\n",
      "32 Q Q <Atom CA>\n",
      "33 F F <Atom CA>\n",
      "34 V V <Atom CA>\n",
      "35 R R <Atom CA>\n",
      "36 F F <Atom CA>\n",
      "37 D D <Atom CA>\n",
      "38 S S <Atom CA>\n",
      "39 D D <Atom CA>\n",
      "40 S S <Atom CA>\n",
      "41 A A <Atom CA>\n",
      "42 S S <Atom CA>\n",
      "43 P P <Atom CA>\n",
      "44 R R <Atom CA>\n",
      "45 M M <Atom CA>\n",
      "46 E E <Atom CA>\n",
      "47 P P <Atom CA>\n",
      "48 R R <Atom CA>\n",
      "49 A A <Atom CA>\n",
      "50 P P <Atom CA>\n",
      "51 W W <Atom CA>\n",
      "52 V V <Atom CA>\n",
      "53 E E <Atom CA>\n",
      "54 Q Q <Atom CA>\n",
      "55 E E <Atom CA>\n",
      "56 G G <Atom CA>\n",
      "57 P P <Atom CA>\n",
      "58 E E <Atom CA>\n",
      "59 Y Y <Atom CA>\n",
      "60 W W <Atom CA>\n",
      "61 E E <Atom CA>\n",
      "62 E E <Atom CA>\n",
      "63 E E <Atom CA>\n",
      "64 T T <Atom CA>\n",
      "65 R R <Atom CA>\n",
      "66 N N <Atom CA>\n",
      "67 T T <Atom CA>\n",
      "68 K K <Atom CA>\n",
      "69 A A <Atom CA>\n",
      "70 H H <Atom CA>\n",
      "71 A A <Atom CA>\n",
      "72 Q Q <Atom CA>\n",
      "73 T T <Atom CA>\n",
      "74 D D <Atom CA>\n",
      "75 R R <Atom CA>\n",
      "76 M M <Atom CA>\n",
      "77 N N <Atom CA>\n",
      "78 L L <Atom CA>\n",
      "79 Q Q <Atom CA>\n",
      "80 T T <Atom CA>\n",
      "81 L L <Atom CA>\n",
      "82 R R <Atom CA>\n",
      "83 G G <Atom CA>\n",
      "84 Y Y <Atom CA>\n",
      "85 Y Y <Atom CA>\n",
      "86 N N <Atom CA>\n",
      "87 Q Q <Atom CA>\n",
      "88 S S <Atom CA>\n",
      "89 E E <Atom CA>\n",
      "90 A A <Atom CA>\n",
      "91 S S <Atom CA>\n",
      "92 S S <Atom CA>\n",
      "93 H H <Atom CA>\n",
      "94 T T <Atom CA>\n",
      "95 L L <Atom CA>\n",
      "96 Q Q <Atom CA>\n",
      "97 W W <Atom CA>\n",
      "98 M M <Atom CA>\n",
      "99 I I <Atom CA>\n",
      "100 G G <Atom CA>\n",
      "101 C C <Atom CA>\n",
      "102 D D <Atom CA>\n",
      "103 L L <Atom CA>\n",
      "104 G G <Atom CA>\n",
      "105 S S <Atom CA>\n",
      "106 D D <Atom CA>\n",
      "107 G G <Atom CA>\n",
      "108 R R <Atom CA>\n",
      "109 L L <Atom CA>\n",
      "110 I I <Atom CA>\n",
      "111 R R <Atom CA>\n",
      "112 G G <Atom CA>\n",
      "113 Y Y <Atom CA>\n",
      "114 E E <Atom CA>\n",
      "115 R R <Atom CA>\n",
      "116 Y Y <Atom CA>\n",
      "117 A A <Atom CA>\n",
      "118 Y Y <Atom CA>\n",
      "119 D D <Atom CA>\n",
      "120 G G <Atom CA>\n",
      "121 K K <Atom CA>\n",
      "122 D D <Atom CA>\n",
      "123 Y Y <Atom CA>\n",
      "124 L L <Atom CA>\n",
      "125 A A <Atom CA>\n",
      "126 L L <Atom CA>\n",
      "127 N N <Atom CA>\n",
      "128 E E <Atom CA>\n",
      "129 D D <Atom CA>\n",
      "130 L L <Atom CA>\n",
      "131 R R <Atom CA>\n",
      "132 S S <Atom CA>\n",
      "133 W W <Atom CA>\n",
      "134 T T <Atom CA>\n",
      "135 A A <Atom CA>\n",
      "136 A A <Atom CA>\n",
      "137 D D <Atom CA>\n",
      "138 T T <Atom CA>\n",
      "139 A A <Atom CA>\n",
      "140 A A <Atom CA>\n",
      "141 Q Q <Atom CA>\n",
      "142 I I <Atom CA>\n",
      "143 S S <Atom CA>\n",
      "144 K K <Atom CA>\n",
      "145 R R <Atom CA>\n",
      "146 K K <Atom CA>\n",
      "147 C C <Atom CA>\n",
      "148 E E <Atom CA>\n",
      "149 A A <Atom CA>\n",
      "150 A A <Atom CA>\n",
      "151 N N <Atom CA>\n",
      "152 V V <Atom CA>\n",
      "153 A A <Atom CA>\n",
      "154 E E <Atom CA>\n",
      "155 Q Q <Atom CA>\n",
      "156 R R <Atom CA>\n",
      "157 R R <Atom CA>\n",
      "158 A A <Atom CA>\n",
      "159 Y Y <Atom CA>\n",
      "160 L L <Atom CA>\n",
      "161 E E <Atom CA>\n",
      "162 G G <Atom CA>\n",
      "163 T T <Atom CA>\n",
      "164 C C <Atom CA>\n",
      "165 V V <Atom CA>\n",
      "166 E E <Atom CA>\n",
      "167 W W <Atom CA>\n",
      "168 L L <Atom CA>\n",
      "169 H H <Atom CA>\n",
      "170 R R <Atom CA>\n",
      "171 Y Y <Atom CA>\n",
      "172 L L <Atom CA>\n",
      "173 E E <Atom CA>\n",
      "174 N N <Atom CA>\n",
      "175 G G <Atom CA>\n",
      "176 K K <Atom CA>\n",
      "177 E E <Atom CA>\n",
      "178 M M <Atom CA>\n",
      "179 L L <Atom CA>\n",
      "180 Q Q <Atom CA>\n",
      "181 R R <Atom CA>\n",
      "182 A A <Atom CA>\n",
      "183 D D <Atom CA>\n",
      "184 P P <Atom CA>\n",
      "185 P P <Atom CA>\n",
      "186 K K <Atom CA>\n",
      "187 T T <Atom CA>\n",
      "188 H H <Atom CA>\n",
      "189 V V <Atom CA>\n",
      "190 T T <Atom CA>\n",
      "191 H H <Atom CA>\n",
      "192 H H <Atom CA>\n",
      "193 P P <Atom CA>\n",
      "194 V V <Atom CA>\n",
      "195 F F <Atom CA>\n",
      "196 D D <Atom CA>\n",
      "197 Y Y <Atom CA>\n",
      "198 E E <Atom CA>\n",
      "199 A A <Atom CA>\n",
      "200 T T <Atom CA>\n",
      "201 L L <Atom CA>\n",
      "202 R R <Atom CA>\n",
      "203 C C <Atom CA>\n",
      "204 W W <Atom CA>\n",
      "205 A A <Atom CA>\n",
      "206 L L <Atom CA>\n",
      "207 G G <Atom CA>\n",
      "208 F F <Atom CA>\n",
      "209 Y Y <Atom CA>\n",
      "210 P P <Atom CA>\n",
      "211 A A <Atom CA>\n",
      "212 E E <Atom CA>\n",
      "213 I I <Atom CA>\n",
      "214 I I <Atom CA>\n",
      "215 L L <Atom CA>\n",
      "216 T T <Atom CA>\n",
      "217 W W <Atom CA>\n",
      "218 Q Q <Atom CA>\n",
      "219 R R <Atom CA>\n",
      "220 D D <Atom CA>\n",
      "221 G G <Atom CA>\n",
      "222 E E <Atom CA>\n",
      "223 D D <Atom CA>\n",
      "224 Q Q <Atom CA>\n",
      "225 T T <Atom CA>\n",
      "226 Q Q <Atom CA>\n",
      "227 D D <Atom CA>\n",
      "228 V V <Atom CA>\n",
      "229 E E <Atom CA>\n",
      "230 L L <Atom CA>\n",
      "231 V V <Atom CA>\n",
      "232 E E <Atom CA>\n",
      "233 T T <Atom CA>\n",
      "234 R R <Atom CA>\n",
      "235 P P <Atom CA>\n",
      "236 A A <Atom CA>\n",
      "237 G G <Atom CA>\n",
      "238 D D <Atom CA>\n",
      "239 G G <Atom CA>\n",
      "240 T T <Atom CA>\n",
      "241 F F <Atom CA>\n",
      "242 Q Q <Atom CA>\n",
      "243 K K <Atom CA>\n",
      "244 W W <Atom CA>\n",
      "245 A A <Atom CA>\n",
      "246 A A <Atom CA>\n",
      "247 V V <Atom CA>\n",
      "248 V V <Atom CA>\n",
      "249 V V <Atom CA>\n",
      "250 P P <Atom CA>\n",
      "251 S S <Atom CA>\n",
      "252 G G <Atom CA>\n",
      "253 E E <Atom CA>\n",
      "254 E E <Atom CA>\n",
      "255 Q Q <Atom CA>\n",
      "256 R R <Atom CA>\n",
      "257 Y Y <Atom CA>\n",
      "258 T T <Atom CA>\n",
      "259 C C <Atom CA>\n",
      "260 H H <Atom CA>\n",
      "261 V V <Atom CA>\n",
      "262 Q Q <Atom CA>\n",
      "263 H H <Atom CA>\n",
      "264 E E <Atom CA>\n",
      "265 G G <Atom CA>\n",
      "266 L L <Atom CA>\n",
      "267 P P <Atom CA>\n",
      "268 E E <Atom CA>\n",
      "269 P P <Atom CA>\n",
      "270 L L <Atom CA>\n",
      "271 M M <Atom CA>\n",
      "272 L L <Atom CA>\n",
      "273 R R <Atom CA>\n",
      "274 W W <Atom CA>\n",
      "275 K K <Atom CA>\n",
      "<Chain id=B>\n",
      "275\n",
      "0 M M <Atom CA>\n",
      "1 I I <Atom CA>\n",
      "2 Q Q <Atom CA>\n",
      "3 R R <Atom CA>\n",
      "4 T T <Atom CA>\n",
      "5 P P <Atom CA>\n",
      "6 K K <Atom CA>\n",
      "7 I I <Atom CA>\n",
      "8 Q Q <Atom CA>\n",
      "9 V V <Atom CA>\n",
      "10 Y Y <Atom CA>\n",
      "11 S S <Atom CA>\n",
      "12 R R <Atom CA>\n",
      "13 H H <Atom CA>\n",
      "14 P P <Atom CA>\n",
      "15 A A <Atom CA>\n",
      "16 E E <Atom CA>\n",
      "17 N N <Atom CA>\n",
      "18 G G <Atom CA>\n",
      "19 K K <Atom CA>\n",
      "20 S S <Atom CA>\n",
      "21 N N <Atom CA>\n",
      "22 F F <Atom CA>\n",
      "23 L L <Atom CA>\n",
      "24 N N <Atom CA>\n",
      "25 C C <Atom CA>\n",
      "26 Y Y <Atom CA>\n",
      "27 V V <Atom CA>\n",
      "28 S S <Atom CA>\n",
      "29 G G <Atom CA>\n",
      "30 F F <Atom CA>\n",
      "31 H H <Atom CA>\n",
      "32 P P <Atom CA>\n",
      "33 S S <Atom CA>\n",
      "34 D D <Atom CA>\n",
      "35 I I <Atom CA>\n",
      "36 E E <Atom CA>\n",
      "37 V V <Atom CA>\n",
      "38 D D <Atom CA>\n",
      "39 L L <Atom CA>\n",
      "40 L L <Atom CA>\n",
      "41 K K <Atom CA>\n",
      "42 N N <Atom CA>\n",
      "43 G G <Atom CA>\n",
      "44 E E <Atom CA>\n",
      "45 R R <Atom CA>\n",
      "46 I I <Atom CA>\n",
      "47 E E <Atom CA>\n",
      "48 K K <Atom CA>\n",
      "49 V V <Atom CA>\n",
      "50 E E <Atom CA>\n",
      "51 H H <Atom CA>\n",
      "52 S S <Atom CA>\n",
      "53 D D <Atom CA>\n",
      "54 L L <Atom CA>\n",
      "55 S S <Atom CA>\n",
      "56 F F <Atom CA>\n",
      "57 S S <Atom CA>\n",
      "58 K K <Atom CA>\n",
      "59 D D <Atom CA>\n",
      "60 W W <Atom CA>\n",
      "61 S S <Atom CA>\n",
      "62 F F <Atom CA>\n",
      "63 Y Y <Atom CA>\n",
      "64 L L <Atom CA>\n",
      "65 L L <Atom CA>\n",
      "66 Y Y <Atom CA>\n",
      "67 Y Y <Atom CA>\n",
      "68 T T <Atom CA>\n",
      "69 E E <Atom CA>\n",
      "70 F F <Atom CA>\n",
      "71 T T <Atom CA>\n",
      "72 P P <Atom CA>\n",
      "73 T T <Atom CA>\n",
      "74 E E <Atom CA>\n",
      "75 K K <Atom CA>\n",
      "76 D D <Atom CA>\n",
      "77 E E <Atom CA>\n",
      "78 Y Y <Atom CA>\n",
      "79 A A <Atom CA>\n",
      "80 C C <Atom CA>\n",
      "81 R R <Atom CA>\n",
      "82 V V <Atom CA>\n",
      "83 N N <Atom CA>\n",
      "84 H H <Atom CA>\n",
      "85 V V <Atom CA>\n",
      "86 T T <Atom CA>\n",
      "87 L L <Atom CA>\n",
      "88 S S <Atom CA>\n",
      "89 Q Q <Atom CA>\n",
      "90 P P <Atom CA>\n",
      "91 K K <Atom CA>\n",
      "92 I I <Atom CA>\n",
      "93 V V <Atom CA>\n",
      "94 K K <Atom CA>\n",
      "95 W W <Atom CA>\n",
      "96 D D <Atom CA>\n",
      "97 R R <Atom CA>\n",
      "98 D D <Atom CA>\n",
      "99 M M <Atom CA>\n",
      "<Chain id=P>\n",
      "375\n",
      "1 K K <Atom CA>\n",
      "2 G G <Atom CA>\n",
      "3 P P <Atom CA>\n",
      "4 P P <Atom CA>\n",
      "5 A A <Atom CA>\n",
      "6 A A <Atom CA>\n",
      "7 L L <Atom CA>\n",
      "8 T T <Atom CA>\n",
      "9 L L <Atom CA>\n"
     ]
    }
   ],
   "source": [
    "## Here I am double checking amino acid signatures from PDB chain to the sequence used in embedding generation\n",
    "\n",
    "residue_position = 0\n",
    "\n",
    "for i in protein_model:\n",
    "    print(i)\n",
    "    print(residue_position)\n",
    "    for residue in i:\n",
    "        try:\n",
    "            print(residue.id[1], simple_aa(residue.resname),protein_sequence[residue_position], residue['CA'])\n",
    "            residue_position+=1\n",
    "        except:\n",
    "            continue\n",
    "#         print(residue1.id[1], residue1.resname)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "601a6af8",
   "metadata": {},
   "source": [
    "## Calculating Expected Contact Sites from PDB structure file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e3318992",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Turned off warnings for discontinuous data structures\n",
    "\n",
    "warnings.simplefilter('ignore', PDBConstructionWarning)\n",
    "\n",
    "# Load the structure from locally saved file\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(pdb_id, f\"{pdb_filename}\")\n",
    "\n",
    "# Extract desired protein structure from PDB structure (typically only 1 structure to choose from)\n",
    "protein_structure = structure[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1d83faae",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Choose protein chain to compare amino acid residues\n",
    "# chain = protein_structure['A']\n",
    "\n",
    "# Checking how many amino acids are compared\n",
    "count = 0\n",
    "\n",
    "for chain in protein_structure:\n",
    "    print(chain)\n",
    "    for residue1 in chain:\n",
    "        for residue2 in chain:\n",
    "            if residue1 != residue2:\n",
    "                # compute distance between CA atoms\n",
    "                try:\n",
    "                    distance = abs(residue1['CA'] - residue2['CA'])\n",
    "                except KeyError:\n",
    "    #                 print(f\"No CA for {residue2.id[1], residue2.resname}\")\n",
    "                    continue\n",
    "                if distance < 5:\n",
    "                    if residue1.id[1] - residue2.id[1] > 2 or residue2.id[1] - residue1.id[1] > 2 :\n",
    "                        print(residue1.id[1], residue1.resname, residue2.id[1], residue2.resname, distance)\n",
    "                        count+=1\n",
    "                    else:\n",
    "                        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19331483",
   "metadata": {},
   "source": [
    "## Initialize Contact Dictionaries for Training Random Forest Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "2187318f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total contacts: 200\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Assuming protein_structure and protein_sequence are defined\n",
    "\n",
    "contacts = defaultdict(dict)\n",
    "\n",
    "# Mapping of chain IDs to starting index\n",
    "chain_to_index = {'A': -1, 'B': 275, 'P': 375}\n",
    "\n",
    "count = 0\n",
    "\n",
    "\n",
    "### Contacts from the SAME chain are only being considered\n",
    "\n",
    "for chain in protein_structure:\n",
    "#     print(chain.id)\n",
    "    index = chain_to_index.get(chain.id, 0)  # Default to 0 if chain ID not found\n",
    "\n",
    "    for residue1 in chain:\n",
    "        for residue2 in chain:\n",
    "            if residue1 != residue2:\n",
    "                try:\n",
    "                    # Calculate Alpha-Carbon Distance\n",
    "                    distance = abs(residue1['CA'] - residue2['CA'])\n",
    "                except KeyError:\n",
    "                    continue\n",
    "\n",
    "                # Calculating Distance of Amino Acids in Sequence\n",
    "                diff = abs(residue1.id[1] - residue2.id[1])\n",
    "                if diff > 2:\n",
    "                    res1_index = residue1.id[1] + index\n",
    "                    res2_index = residue2.id[1] + index\n",
    "\n",
    "                    # Ensure indices are within the bounds of protein_sequence\n",
    "                    if 0 <= res1_index < len(protein_sequence) and 0 <= res2_index < len(protein_sequence):\n",
    "                        contact = distance < 5\n",
    "                        contacts[(res1_index, res2_index)] = {\n",
    "                            'aa1': protein_sequence[res1_index],\n",
    "                            'aa2': protein_sequence[res2_index], \n",
    "                            'dist': distance,\n",
    "                            'contact': contact\n",
    "                        }\n",
    "                        if contact:\n",
    "                            count += 1\n",
    "#                         print(res1_index, residue1.resname, res2_index, residue2.resname, distance, protein_sequence[res1_index], protein_sequence[res2_index], contact)\n",
    "\n",
    "print(f\"Total contacts: {count}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27249c6f",
   "metadata": {},
   "source": [
    "## Implementing Random Forest Classifier (Currently for only 1 sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "7eff13ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "c2bbfac6",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "[tensor(0.0011), tensor(0.0031), tensor(0.0022), tensor(0.0032), tensor(0.0035)]\n"
     ]
    }
   ],
   "source": [
    "# Extract attentions from all heads and layers for given amino acid residues\n",
    "\n",
    "def get_x_y(res1, res2):\n",
    "    vectors = []\n",
    "    in_contact = contacts[(res1,res2)]['contact']\n",
    "    for layer in range(0,33):\n",
    "        for head in range(0,20):\n",
    "            vectors.append(results['attentions'][0][layer][head][res1][res2])\n",
    "\n",
    "    return vectors, in_contact\n",
    "\n",
    "vectors, in_contact = get_x_y(1,103)\n",
    "print(in_contact)\n",
    "print(vectors[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "6efca0b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set X to extracted attentions from layers and heads\n",
    "# Set Y to True/False if the amino acids are in contact\n",
    "\n",
    "X = []\n",
    "y = []\n",
    "\n",
    "for i in list(contacts.keys())[:1500]:\n",
    "    vectors, in_contact = get_x_y(i[0],i[1])\n",
    "    X.append(vectors)\n",
    "    y.append(in_contact)\n",
    "    \n",
    "X = np.array(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "26487caf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-4 {color: black;}#sk-container-id-4 pre{padding: 0;}#sk-container-id-4 div.sk-toggleable {background-color: white;}#sk-container-id-4 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-4 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-4 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-4 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-4 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-4 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-4 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-4 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-4 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-4 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-4 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-4 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-4 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-4 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-4 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-4 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-4 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-4 div.sk-item {position: relative;z-index: 1;}#sk-container-id-4 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-4 div.sk-item::before, #sk-container-id-4 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-4 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-4 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-4 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-4 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-4 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-4 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-4 div.sk-label-container {text-align: center;}#sk-container-id-4 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-4 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-4\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(max_depth=2, random_state=0)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-4\" type=\"checkbox\" checked><label for=\"sk-estimator-id-4\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(max_depth=2, random_state=0)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(max_depth=2, random_state=0)"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fit classifier\n",
    "\n",
    "clf = RandomForestClassifier(max_depth=2, random_state=0)\n",
    "clf.fit(X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "7d6fb4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Find X vectors that are supposed to be True\n",
    "\n",
    "# count = 0\n",
    "\n",
    "# for vec,cont in zip(X,y):\n",
    "#     count+=1\n",
    "#     if cont == True:\n",
    "#         print(count)\n",
    "#         print(vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "53250908",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aa1': 'S', 'aa2': 'G', 'dist': 4.9678664, 'contact': True}"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contacts[(1,103)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "d259122d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False])"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing prediction on known True contact\n",
    "\n",
    "vectors, in_contact = get_x_y(1, 103)\n",
    "# vectors = X[372]\n",
    "vectors = np.array(vectors).reshape(1, -1)  # Reshape for a single sample prediction\n",
    "prediction = clf.predict(vectors)\n",
    "prediction"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
