{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1cbea00c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ESM2(\n",
      "  (embed_tokens): Embedding(33, 1280, padding_idx=1)\n",
      "  (layers): ModuleList(\n",
      "    (0-32): 33 x TransformerLayer(\n",
      "      (self_attn): MultiheadAttention(\n",
      "        (k_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
      "        (v_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
      "        (q_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
      "        (out_proj): Linear(in_features=1280, out_features=1280, bias=True)\n",
      "        (rot_emb): RotaryEmbedding()\n",
      "      )\n",
      "      (self_attn_layer_norm): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
      "      (fc1): Linear(in_features=1280, out_features=5120, bias=True)\n",
      "      (fc2): Linear(in_features=5120, out_features=1280, bias=True)\n",
      "      (final_layer_norm): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
      "    )\n",
      "  )\n",
      "  (contact_head): ContactPredictionHead(\n",
      "    (regression): Linear(in_features=660, out_features=1, bias=True)\n",
      "    (activation): Sigmoid()\n",
      "  )\n",
      "  (emb_layer_norm_after): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
      "  (lm_head): RobertaLMHead(\n",
      "    (dense): Linear(in_features=1280, out_features=1280, bias=True)\n",
      "    (layer_norm): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)\n",
      "  )\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "from extract_contacts_and_attentions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e6294372",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter('ignore', PDBConstructionWarning)\n",
    "parser = PDBParser()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b7780d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_sequences(pdb_id):  \n",
    "    # Parse pdb file and save as structure. The pdb file is where we are getting CA coordinates from.\n",
    "    structure = parser.get_structure(pdb_id, f\"{structure_dir+pdb_id}.pdb\")\n",
    "\n",
    "    # Extract desired protein structure from PDB structure (typically only 1 structure to choose from)\n",
    "    protein_structure = structure[0]\n",
    "\n",
    "    residue_position = 0\n",
    "    mismatches = 0\n",
    "#     print(pdb_id)\n",
    "    if 'A' in protein_structure:\n",
    "        for residue in protein_structure['A']:\n",
    "            if 'CA' in residue:\n",
    "                if residue_position < len(protein_data[pdb_id]):\n",
    "                    if simple_aa(residue.resname) != protein_data[pdb_id][residue_position]:\n",
    "#                         print(residue.id[1], simple_aa(residue.resname), protein_data[pdb_id][residue_position])\n",
    "                        mismatches+=1\n",
    "            residue_position+=1\n",
    "        if mismatches == 0:\n",
    "#             same_sequence_ids.append(pdb_id)\n",
    "            return pdb_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96663a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_casp_pdb_seqs(protein_data):\n",
    "    same_sequence_ids = []\n",
    "    iterations = 0\n",
    "    for pdb_id, protein_sequence in list(protein_data.items()):\n",
    "#         print(\"Iterations: \", iterations)\n",
    "#         print('Sequence ID: ', pdb_id)\n",
    "#         check_sequences(pdb_id)\n",
    "#         print('No mismatches in pdb sequence and pulled sequence: ', check_sequences(pdb_id))\n",
    "        if check_sequences(pdb_id) == None:\n",
    "            continue\n",
    "        else:\n",
    "            same_sequence_ids.append(pdb_id)\n",
    "        iterations+=1\n",
    "    return same_sequence_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "15d32be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_contact_sites(pdb_id, in_contact_sites, non_contact_sites, subset_non_contact_sites):\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure(pdb_id, f\"{structure_dir}/{pdb_id}.pdb\")  # Ensure correct path joining\n",
    "    protein_structure = structure[0]\n",
    "    chain = protein_structure['A']\n",
    "\n",
    "    # Initialize count variable\n",
    "    count = 0\n",
    "\n",
    "    for i, residue1 in enumerate(chain):\n",
    "        for j, residue2 in enumerate(chain):\n",
    "            if i <= j:\n",
    "                continue # Avoids redundant comparisons and self-comparison\n",
    "            if residue1.id[1] > len(protein_data[pdb_id]) or residue2.id[1] > len(protein_data[pdb_id]):\n",
    "                continue\n",
    "            try:\n",
    "                distance = abs(residue1['CA'] - residue2['CA'])\n",
    "            except KeyError:\n",
    "                continue\n",
    "            if distance < 5:\n",
    "                if abs(residue1.id[1] - residue2.id[1]) > 2:\n",
    "    #                 print(residue1.id[1], residue1.resname, residue2.id[1], residue2.resname, distance)\n",
    "                    in_contact_sites[pdb_id].append({\n",
    "                        'res_1': residue1.id[1], \n",
    "                        'res_2': residue2.id[1], \n",
    "                        'sig_1': simple_aa(residue1.resname), \n",
    "                        'sig_2': simple_aa(residue2.resname), \n",
    "                        'dist': distance,\n",
    "                        'in_contact': True\n",
    "                    })\n",
    "                    count += 1\n",
    "            else:\n",
    "                if abs(residue1.id[1] - residue2.id[1]) > 2:\n",
    "                    non_contact_sites[pdb_id].append({\n",
    "                        'res_1': residue1.id[1], \n",
    "                        'res_2': residue2.id[1], \n",
    "                        'sig_1': simple_aa(residue1.resname), \n",
    "                        'sig_2': simple_aa(residue2.resname), \n",
    "                        'dist': distance,\n",
    "                        'in_contact': False\n",
    "                    })\n",
    "\n",
    "    if non_contact_sites[pdb_id]:\n",
    "        subset_non_contact_sites[pdb_id] = random.sample(non_contact_sites[pdb_id], min(len(non_contact_sites[pdb_id]), len(in_contact_sites[pdb_id])))\n",
    "\n",
    "    # Optionally print or process the results\n",
    "    return f\"Total contacts found {pdb_id}: {count}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ceeb8532",
   "metadata": {},
   "outputs": [],
   "source": [
    "def contacts_per_pdb(same_sequence_ids):\n",
    "\n",
    "    in_contact_sites = defaultdict(list)\n",
    "    non_contact_sites = defaultdict(list)\n",
    "    subset_non_contact_sites = defaultdict(list)\n",
    "\n",
    "    iterations = 0\n",
    "\n",
    "    for pdb_id in same_sequence_ids:\n",
    "        calc_contact_sites(pdb_id, in_contact_sites, non_contact_sites, subset_non_contact_sites)\n",
    "#         print(calc_contact_sites(pdb_id, in_contact_sites, non_contact_sites, subset_non_contact_sites))\n",
    "#         print(len(in_contact_sites[pdb_id]), len(subset_non_contact_sites[pdb_id]))\n",
    "#         print(\"Iterations: \", iterations)\n",
    "        iterations+=1\n",
    "        \n",
    "    return in_contact_sites, non_contact_sites, subset_non_contact_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d9139ddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize contact_data as a defaultdict of lists\n",
    "\n",
    "def generate_contact_data(in_contact_sites, subset_non_contact_sites):\n",
    "\n",
    "    contact_data = defaultdict(list)\n",
    "\n",
    "    # Add data from in_contact_sites\n",
    "    for pdb_id, contacts in in_contact_sites.items():\n",
    "        for contact in contacts:\n",
    "            contact_data[pdb_id].append({\n",
    "                'res_1': contact['res_1'],\n",
    "                'res_2': contact['res_2'],\n",
    "                'sig_1': contact['sig_1'],\n",
    "                'sig_2': contact['sig_2'],\n",
    "                'dist': contact['dist'],\n",
    "                'in_contact': contact['in_contact']\n",
    "            })\n",
    "\n",
    "    # Add data from subset_non_contact_sites\n",
    "    for pdb_id, non_contacts in subset_non_contact_sites.items():\n",
    "        for non_contact in non_contacts:\n",
    "            contact_data[pdb_id].append({\n",
    "                'res_1': non_contact['res_1'],\n",
    "                'res_2': non_contact['res_2'],\n",
    "                'sig_1': non_contact['sig_1'],\n",
    "                'sig_2': non_contact['sig_2'],\n",
    "                'dist': non_contact['dist'],\n",
    "                'in_contact': non_contact['in_contact']\n",
    "            })\n",
    "    return contact_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1093e09e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Index protein sequence as sequence 0 (next sequence would be indexed as 1)\n",
    "\n",
    "def generate_embeddings(pdb_id):\n",
    "    protein_sequence = protein_data[pdb_id]\n",
    "    esm_input_data = [(0, protein_sequence)]\n",
    "    # print('Data: ', esm_input_data, '\\n')\n",
    "\n",
    "    # Prepare variables to input sequence into ESM-2 model \n",
    "    batch_converter = alphabet.get_batch_converter()\n",
    "    batch_labels, batch_strs, batch_tokens = batch_converter(esm_input_data)\n",
    "    batch_tokens = batch_tokens.cuda() if torch.cuda.is_available() else batch_tokens\n",
    "\n",
    "    # print('batch_tokens: ', '\\n\\n', batch_tokens, '\\n')\n",
    "\n",
    "    # 4. Input prepared sequence information into model and output as results (contact predictions are included in embedding)\n",
    "    with torch.no_grad():\n",
    "        results = model(batch_tokens, repr_layers=[33], return_contacts=True)\n",
    "\n",
    "    return results['attentions']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9fd8cc48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract attentions from all heads and layers for given amino acid residues\n",
    "\n",
    "def get_x_y(attention_data, res_1, res_2):\n",
    "    vectors = []\n",
    "    for layer in range(0,33):\n",
    "        for head in range(0,20):\n",
    "            vectors.append(attention_data[0][layer][head][res_1][res_2])\n",
    "\n",
    "    return vectors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b1d582e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def output_x_y(sequence_ids, contact_data):\n",
    "    X = []\n",
    "    y = []\n",
    "    iterations = 0\n",
    "\n",
    "    for pdb_id in sequence_ids:\n",
    "        structure = parser.get_structure(pdb_id, f\"{structure_dir}/{pdb_id}.pdb\")  # Ensure correct path joining\n",
    "        protein_structure = structure[0]\n",
    "        chain = protein_structure['A']\n",
    "        first_residue = list(chain.get_residues())[0].id[1]\n",
    "        print('Iteration: ', iterations)\n",
    "        iterations+=1\n",
    "        if first_residue == 1:\n",
    "            attention_data = generate_embeddings(pdb_id)\n",
    "            for i in contact_data[pdb_id]:\n",
    "                    X.append(get_x_y(attention_data, i['res_1'], i['res_2']))\n",
    "                    y.append(i['in_contact'])\n",
    "            else:\n",
    "                continue\n",
    "    return X,y\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e38f05d4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get sequences from CASP7 file\n",
    "prot_data_dict = parse_casp7_file(casp_95)\n",
    "\n",
    "protein_id_counts = Counter(protein_id.split('_')[0] for protein_id in prot_data_dict)\n",
    "single_occurence_ids = [protein_id for protein_id, count in protein_id_counts.items() if count == 1]\n",
    "\n",
    "# protein_data = generate_fastas(single_occurence_ids)\n",
    "protein_data = load_fastas(fasta_dir)\n",
    "same_sequence_ids = check_casp_pdb_seqs(protein_data)\n",
    "in_contact_sites, non_contact_sites, subset_non_contact_sites = contacts_per_pdb(same_sequence_ids)\n",
    "contact_data = generate_contact_data(in_contact_sites, subset_non_contact_sites)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7c590674",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1UE8', '1F8E', '1J5B', '1P0Z', '1N3K']\n"
     ]
    }
   ],
   "source": [
    "seed_value = 67\n",
    "random.seed(seed_value)\n",
    "n_sequences = 300\n",
    "\n",
    "# sequence_ids = random.sample(same_sequence_ids, n_sequences)\n",
    "sequence_ids = same_sequence_ids\n",
    "print(sequence_ids[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8dc97a2",
   "metadata": {},
   "source": [
    "## Linear SVC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a67d48bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.svm import LinearSVC\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d77b6c77",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration:  0\n",
      "Iteration:  1\n",
      "Iteration:  2\n",
      "Iteration:  3\n",
      "Iteration:  4\n",
      "Iteration:  5\n",
      "Iteration:  6\n",
      "Iteration:  7\n",
      "Iteration:  8\n",
      "Iteration:  9\n",
      "Iteration:  10\n",
      "Iteration:  11\n",
      "Iteration:  12\n",
      "Iteration:  13\n",
      "Iteration:  14\n",
      "Iteration:  15\n",
      "Iteration:  16\n",
      "Iteration:  17\n",
      "Iteration:  18\n",
      "Iteration:  19\n",
      "Iteration:  20\n",
      "Iteration:  21\n",
      "Iteration:  22\n",
      "Iteration:  23\n",
      "Iteration:  24\n",
      "Iteration:  25\n",
      "Iteration:  26\n",
      "Iteration:  27\n",
      "Iteration:  28\n",
      "Iteration:  29\n",
      "Iteration:  30\n",
      "Iteration:  31\n",
      "Iteration:  32\n",
      "Iteration:  33\n",
      "Iteration:  34\n",
      "Iteration:  35\n",
      "Iteration:  36\n",
      "Iteration:  37\n",
      "Iteration:  38\n",
      "Iteration:  39\n",
      "Iteration:  40\n",
      "Iteration:  41\n",
      "Iteration:  42\n",
      "Iteration:  43\n",
      "Iteration:  44\n",
      "Iteration:  45\n",
      "Iteration:  46\n",
      "Iteration:  47\n",
      "Iteration:  48\n",
      "Iteration:  49\n",
      "Iteration:  50\n",
      "Iteration:  51\n",
      "Iteration:  52\n",
      "Iteration:  53\n",
      "Iteration:  54\n",
      "Iteration:  55\n",
      "Iteration:  56\n",
      "Iteration:  57\n",
      "Iteration:  58\n",
      "Iteration:  59\n",
      "Iteration:  60\n",
      "Iteration:  61\n",
      "Iteration:  62\n",
      "Iteration:  63\n",
      "Iteration:  64\n",
      "Iteration:  65\n",
      "Iteration:  66\n",
      "Iteration:  67\n",
      "Iteration:  68\n",
      "Iteration:  69\n",
      "Iteration:  70\n",
      "Iteration:  71\n",
      "Iteration:  72\n",
      "Iteration:  73\n",
      "Iteration:  74\n",
      "Iteration:  75\n",
      "Iteration:  76\n",
      "Iteration:  77\n",
      "Iteration:  78\n",
      "Iteration:  79\n",
      "Iteration:  80\n",
      "Iteration:  81\n",
      "Iteration:  82\n",
      "Iteration:  83\n",
      "Iteration:  84\n",
      "Iteration:  85\n",
      "Iteration:  86\n",
      "Iteration:  87\n",
      "Iteration:  88\n",
      "Iteration:  89\n",
      "Iteration:  90\n",
      "Iteration:  91\n",
      "Iteration:  92\n",
      "Iteration:  93\n",
      "Iteration:  94\n",
      "Iteration:  95\n",
      "Iteration:  96\n",
      "Iteration:  97\n",
      "Iteration:  98\n",
      "Iteration:  99\n",
      "Iteration:  100\n",
      "Iteration:  101\n",
      "Iteration:  102\n",
      "Iteration:  103\n",
      "Iteration:  104\n",
      "Iteration:  105\n",
      "Iteration:  106\n",
      "Iteration:  107\n",
      "Iteration:  108\n",
      "Iteration:  109\n",
      "Iteration:  110\n",
      "Iteration:  111\n",
      "Iteration:  112\n",
      "Iteration:  113\n",
      "Iteration:  114\n",
      "Iteration:  115\n",
      "Iteration:  116\n",
      "Iteration:  117\n",
      "Iteration:  118\n",
      "Iteration:  119\n",
      "Iteration:  120\n",
      "Iteration:  121\n",
      "Iteration:  122\n",
      "Iteration:  123\n",
      "Iteration:  124\n",
      "Iteration:  125\n",
      "Iteration:  126\n",
      "Iteration:  127\n",
      "Iteration:  128\n",
      "Iteration:  129\n",
      "Iteration:  130\n",
      "Iteration:  131\n",
      "Iteration:  132\n",
      "Iteration:  133\n",
      "Iteration:  134\n",
      "Iteration:  135\n",
      "Iteration:  136\n",
      "Iteration:  137\n",
      "Iteration:  138\n",
      "Iteration:  139\n",
      "Iteration:  140\n",
      "Iteration:  141\n",
      "Iteration:  142\n",
      "Iteration:  143\n",
      "Iteration:  144\n",
      "Iteration:  145\n",
      "Iteration:  146\n",
      "Iteration:  147\n",
      "Iteration:  148\n",
      "Iteration:  149\n",
      "Iteration:  150\n",
      "Iteration:  151\n",
      "Iteration:  152\n",
      "Iteration:  153\n",
      "Iteration:  154\n",
      "Iteration:  155\n",
      "Iteration:  156\n",
      "Iteration:  157\n",
      "Iteration:  158\n",
      "Iteration:  159\n",
      "Iteration:  160\n",
      "Iteration:  161\n",
      "Iteration:  162\n",
      "Iteration:  163\n",
      "Iteration:  164\n",
      "Iteration:  165\n",
      "Iteration:  166\n",
      "Iteration:  167\n",
      "Iteration:  168\n",
      "Iteration:  169\n",
      "Iteration:  170\n",
      "Iteration:  171\n",
      "Iteration:  172\n",
      "Iteration:  173\n",
      "Iteration:  174\n",
      "Iteration:  175\n",
      "Iteration:  176\n",
      "Iteration:  177\n",
      "Iteration:  178\n",
      "Iteration:  179\n",
      "Iteration:  180\n",
      "Iteration:  181\n",
      "Iteration:  182\n",
      "Iteration:  183\n",
      "Iteration:  184\n",
      "Iteration:  185\n",
      "Iteration:  186\n",
      "Iteration:  187\n",
      "Iteration:  188\n",
      "Iteration:  189\n",
      "Iteration:  190\n",
      "Iteration:  191\n",
      "Iteration:  192\n",
      "Iteration:  193\n",
      "Iteration:  194\n",
      "Iteration:  195\n",
      "Iteration:  196\n",
      "Iteration:  197\n",
      "Iteration:  198\n",
      "Iteration:  199\n",
      "Iteration:  200\n",
      "Iteration:  201\n",
      "Iteration:  202\n",
      "Iteration:  203\n",
      "Iteration:  204\n",
      "Iteration:  205\n",
      "Iteration:  206\n",
      "Iteration:  207\n",
      "Iteration:  208\n",
      "Iteration:  209\n",
      "Iteration:  210\n",
      "Iteration:  211\n",
      "Iteration:  212\n",
      "Iteration:  213\n",
      "Iteration:  214\n",
      "Iteration:  215\n",
      "Iteration:  216\n",
      "Iteration:  217\n",
      "Iteration:  218\n",
      "Iteration:  219\n",
      "Iteration:  220\n",
      "Iteration:  221\n",
      "Iteration:  222\n",
      "Iteration:  223\n",
      "Iteration:  224\n",
      "Iteration:  225\n",
      "Iteration:  226\n",
      "Iteration:  227\n",
      "Iteration:  228\n",
      "Iteration:  229\n",
      "Iteration:  230\n",
      "Iteration:  231\n",
      "Iteration:  232\n",
      "Iteration:  233\n",
      "Iteration:  234\n",
      "Iteration:  235\n",
      "Iteration:  236\n",
      "Iteration:  237\n",
      "Iteration:  238\n",
      "Iteration:  239\n",
      "Iteration:  240\n",
      "Iteration:  241\n",
      "Iteration:  242\n",
      "Iteration:  243\n",
      "Iteration:  244\n",
      "Iteration:  245\n",
      "Iteration:  246\n",
      "Iteration:  247\n",
      "Iteration:  248\n",
      "Iteration:  249\n",
      "Iteration:  250\n",
      "Iteration:  251\n",
      "Iteration:  252\n",
      "Iteration:  253\n",
      "Iteration:  254\n",
      "Iteration:  255\n",
      "Iteration:  256\n",
      "Iteration:  257\n",
      "Iteration:  258\n",
      "Iteration:  259\n",
      "Iteration:  260\n",
      "Iteration:  261\n",
      "Iteration:  262\n",
      "Iteration:  263\n",
      "Iteration:  264\n",
      "Iteration:  265\n",
      "Iteration:  266\n",
      "Iteration:  267\n",
      "Iteration:  268\n",
      "Iteration:  269\n",
      "Iteration:  270\n",
      "Iteration:  271\n",
      "Iteration:  272\n",
      "Iteration:  273\n",
      "Iteration:  274\n",
      "Iteration:  275\n",
      "Iteration:  276\n",
      "Iteration:  277\n",
      "Iteration:  278\n",
      "Iteration:  279\n",
      "Iteration:  280\n",
      "Iteration:  281\n",
      "Iteration:  282\n",
      "Iteration:  283\n",
      "Iteration:  284\n",
      "Iteration:  285\n",
      "Iteration:  286\n",
      "Iteration:  287\n",
      "Iteration:  288\n",
      "Iteration:  289\n",
      "Iteration:  290\n",
      "Iteration:  291\n",
      "Iteration:  292\n",
      "Iteration:  293\n",
      "Iteration:  294\n",
      "Iteration:  295\n",
      "Iteration:  296\n",
      "Iteration:  297\n",
      "Iteration:  298\n",
      "Iteration:  299\n",
      "Iteration:  300\n",
      "Iteration:  301\n",
      "Iteration:  302\n",
      "Iteration:  303\n",
      "Iteration:  304\n",
      "Iteration:  305\n",
      "Iteration:  306\n",
      "Iteration:  307\n",
      "Iteration:  308\n",
      "Iteration:  309\n",
      "Iteration:  310\n",
      "Iteration:  311\n",
      "Iteration:  312\n",
      "Iteration:  313\n",
      "Iteration:  314\n",
      "Iteration:  315\n",
      "Iteration:  316\n",
      "Iteration:  317\n",
      "Iteration:  318\n",
      "Iteration:  319\n",
      "Iteration:  320\n",
      "Iteration:  321\n",
      "Iteration:  322\n",
      "Iteration:  323\n",
      "Iteration:  324\n",
      "Iteration:  325\n",
      "Iteration:  326\n",
      "Iteration:  327\n",
      "Iteration:  328\n",
      "Iteration:  329\n",
      "Iteration:  330\n",
      "Iteration:  331\n",
      "Iteration:  332\n",
      "Iteration:  333\n",
      "Iteration:  334\n",
      "Iteration:  335\n",
      "Iteration:  336\n",
      "Iteration:  337\n",
      "Iteration:  338\n",
      "Iteration:  339\n",
      "Iteration:  340\n",
      "Iteration:  341\n",
      "Iteration:  342\n",
      "Iteration:  343\n",
      "Iteration:  344\n",
      "Iteration:  345\n",
      "Iteration:  346\n",
      "Iteration:  347\n",
      "Iteration:  348\n",
      "Iteration:  349\n",
      "Iteration:  350\n",
      "Iteration:  351\n",
      "Iteration:  352\n",
      "Iteration:  353\n",
      "Iteration:  354\n",
      "Iteration:  355\n",
      "Iteration:  356\n",
      "Iteration:  357\n",
      "Iteration:  358\n",
      "Iteration:  359\n",
      "Iteration:  360\n",
      "Iteration:  361\n",
      "Iteration:  362\n",
      "Iteration:  363\n",
      "Iteration:  364\n",
      "Iteration:  365\n",
      "Iteration:  366\n",
      "Iteration:  367\n",
      "Iteration:  368\n",
      "Iteration:  369\n",
      "Iteration:  370\n",
      "Iteration:  371\n",
      "Iteration:  372\n",
      "Iteration:  373\n",
      "Iteration:  374\n",
      "Iteration:  375\n",
      "Iteration:  376\n",
      "Iteration:  377\n",
      "Iteration:  378\n",
      "Iteration:  379\n",
      "Iteration:  380\n",
      "Iteration:  381\n",
      "Iteration:  382\n",
      "Iteration:  383\n",
      "Iteration:  384\n",
      "Iteration:  385\n",
      "Iteration:  386\n",
      "Iteration:  387\n",
      "Iteration:  388\n",
      "Iteration:  389\n",
      "Iteration:  390\n",
      "Iteration:  391\n",
      "Iteration:  392\n",
      "Iteration:  393\n",
      "Iteration:  394\n",
      "Iteration:  395\n",
      "Iteration:  396\n",
      "Iteration:  397\n",
      "Iteration:  398\n",
      "Iteration:  399\n",
      "Iteration:  400\n",
      "Iteration:  401\n",
      "Iteration:  402\n",
      "Iteration:  403\n",
      "Iteration:  404\n",
      "Iteration:  405\n",
      "Iteration:  406\n",
      "Iteration:  407\n",
      "Iteration:  408\n",
      "Iteration:  409\n",
      "Iteration:  410\n",
      "Iteration:  411\n",
      "Iteration:  412\n",
      "Iteration:  413\n",
      "Iteration:  414\n",
      "Iteration:  415\n",
      "Iteration:  416\n",
      "Iteration:  417\n",
      "Iteration:  418\n",
      "Iteration:  419\n",
      "Iteration:  420\n",
      "Iteration:  421\n",
      "Iteration:  422\n",
      "Iteration:  423\n",
      "Iteration:  424\n",
      "Iteration:  425\n",
      "Iteration:  426\n",
      "Iteration:  427\n",
      "Iteration:  428\n",
      "Iteration:  429\n",
      "Iteration:  430\n",
      "Iteration:  431\n",
      "Iteration:  432\n",
      "Iteration:  433\n",
      "Iteration:  434\n",
      "Iteration:  435\n",
      "Iteration:  436\n",
      "Iteration:  437\n",
      "Iteration:  438\n",
      "Iteration:  439\n",
      "Iteration:  440\n",
      "Iteration:  441\n",
      "Iteration:  442\n",
      "Iteration:  443\n",
      "Iteration:  444\n",
      "Iteration:  445\n",
      "Iteration:  446\n",
      "Iteration:  447\n",
      "Iteration:  448\n",
      "Iteration:  449\n",
      "Iteration:  450\n",
      "Iteration:  451\n",
      "Iteration:  452\n",
      "Iteration:  453\n",
      "Iteration:  454\n",
      "Iteration:  455\n",
      "Iteration:  456\n",
      "Iteration:  457\n",
      "Iteration:  458\n",
      "Iteration:  459\n",
      "Iteration:  460\n",
      "Iteration:  461\n",
      "Iteration:  462\n",
      "Iteration:  463\n",
      "Iteration:  464\n",
      "Iteration:  465\n",
      "Iteration:  466\n",
      "Iteration:  467\n",
      "Iteration:  468\n",
      "Iteration:  469\n",
      "Iteration:  470\n",
      "Iteration:  471\n",
      "Iteration:  472\n",
      "Iteration:  473\n",
      "Iteration:  474\n",
      "Iteration:  475\n",
      "Iteration:  476\n",
      "Iteration:  477\n",
      "Iteration:  478\n",
      "Iteration:  479\n",
      "Iteration:  480\n",
      "Iteration:  481\n",
      "Iteration:  482\n",
      "Iteration:  483\n",
      "Iteration:  484\n",
      "Iteration:  485\n",
      "Iteration:  486\n",
      "Iteration:  487\n",
      "Iteration:  488\n",
      "Iteration:  489\n",
      "Iteration:  490\n",
      "Iteration:  491\n",
      "Iteration:  492\n",
      "Iteration:  493\n",
      "Iteration:  494\n",
      "Iteration:  495\n",
      "Iteration:  496\n",
      "Iteration:  497\n",
      "Iteration:  498\n",
      "Iteration:  499\n",
      "Iteration:  500\n",
      "Iteration:  501\n",
      "Iteration:  502\n",
      "Iteration:  503\n",
      "Iteration:  504\n",
      "Iteration:  505\n",
      "Iteration:  506\n",
      "Iteration:  507\n",
      "Iteration:  508\n",
      "Iteration:  509\n",
      "Iteration:  510\n",
      "Iteration:  511\n",
      "Iteration:  512\n",
      "Iteration:  513\n",
      "Iteration:  514\n",
      "Iteration:  515\n",
      "Iteration:  516\n",
      "Iteration:  517\n",
      "Iteration:  518\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration:  519\n",
      "Iteration:  520\n",
      "Iteration:  521\n",
      "Iteration:  522\n",
      "Iteration:  523\n",
      "Iteration:  524\n",
      "Iteration:  525\n",
      "Iteration:  526\n",
      "Iteration:  527\n",
      "Iteration:  528\n",
      "Iteration:  529\n",
      "Iteration:  530\n",
      "Iteration:  531\n",
      "Iteration:  532\n",
      "Iteration:  533\n",
      "Iteration:  534\n",
      "Iteration:  535\n",
      "Iteration:  536\n",
      "Iteration:  537\n",
      "Iteration:  538\n",
      "Iteration:  539\n",
      "Iteration:  540\n",
      "Iteration:  541\n",
      "Iteration:  542\n",
      "Iteration:  543\n",
      "Iteration:  544\n"
     ]
    }
   ],
   "source": [
    "X, y = output_x_y(sequence_ids, contact_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a5eec35",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.4, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75f50776",
   "metadata": {},
   "outputs": [],
   "source": [
    "linear_svc = LinearSVC(random_state=55)\n",
    "linear_svc.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04e4300f",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = linear_svc.predict(X_test)\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(f\"Accuracy: {accuracy:.2f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
